enterprise_gateway/services/sessions/kernelsessionmanager.py

```diff
@@ -279,22 +277,48 @@ def persistence_root_default(self):
     def __init__(self, kernel_manager, **kwargs):
         super(FileKernelSessionManager, self).__init__(kernel_manager, **kwargs)
         if self.enable_persistence:
-            self.kernel_session_file = os.path.join(self._get_sessions_loc(), KERNEL_SESSIONS_FILE_NAME)
-            self.log.info("Kernel session persistence location: {}".format(self.kernel_session_file))
+            self.log.info("Kernel session persistence location: {}".format(self._get_sessions_loc()))
 
     def save_sessions(self):
         if self.enable_persistence:
-            with open(self.kernel_session_file, 'w') as fp:
-                json.dump(KernelSessionManager.pre_save_transformation(self._sessions), fp)
-                fp.close()
+            os.rmdir(self._get_sessions_loc())
+            for kernel_id, session in self._sessions.items():
+                if kernel_id is not None:
+                    json_file_name = "".join(['kernel_', kernel_id, '.json'])
+                    kernel_session_file_path = os.path.join(self._get_sessions_loc(), json_file_name)
+                    temp_session = dict()
+                    temp_session[kernel_id] = session
+                    with open(kernel_session_file_path, 'w') as fp:
+                        json.dump(KernelSessionManager.pre_save_transformation(temp_session), fp)
+                        fp.close()
 
     def load_sessions(self):
         if self.enable_persistence:
-            if os.path.exists(self.kernel_session_file):
-                self.log.debug("Loading saved sessions from {}".format(self.kernel_session_file))
-                with open(self.kernel_session_file) as fp:
-                    self._sessions = KernelSessionManager.post_load_transformation(json.load(fp))
+            self._load_sessions()
+
+    def _load_sessions(self, kernel_id=None):
+        if kernel_id is not None:
+            kernel_file = "".join(['kernel_', kernel_id, '.json'])
+            kernel_session_file_path = os.path.join(self._get_sessions_loc(), kernel_file)
+            temp_sessions = self._save_sessions(kernel_session_file_path)
+        else:
+            kernel_session_files = [json_files for json_files in os.listdir(self._get_sessions_loc()) if
+                                    json_files.endswith('.json')]
+
+            temp_sessions = self._save_sessions(kernel_session_files)
+
+        self._sessions = KernelSessionManager.post_load_transformation(temp_sessions)
+
+    def _save_sessions(self, kernel_session_files):
```

Comment:
Not sure this method is necessary.  I think `post_load_transformation()` and `pre_save_transformation()` need to be updated to work on single sessions and then the assignment (on line 310 above) would be an update to `self._sessions` for a specific kernel_id and its loaded (and transformed) session.

Reply:
Again I didn't want to change any base methods, so I built subclass accordingly. Let me make these changes and update the PR.